### ISSUE TITLE
`var` nullability analysis regression

### ISSUE THREAD

**Issue by davidmilligan** [dotnet/roslyn#52925]
**Version Used**: 
Microsoft.Net.Compilers 3.9.0

**Steps to Reproduce**:

```cs
#nullable enable

class C<T> where T : notnull
{
    void M(T initial, System.Func<T, T?> selector)
    {
        // ⚠ CS8600 Converting null literal or possible null value to non-nullable type.
        //                                                     ↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓
        for (var current = initial; current != null; current = selector(current))
        {
        }
    }
}
```

**Expected Behavior**:

No CS8600 warning

**Actual Behavior**:

CS8600 warning

**Notes**:
This code snippet previous compiled without warning on 3.8.0, so this appears to be a regression. It also only happens with the constraint `T : notnull`. `T : class` for example does not give the warning

**Comment by jcouv** [comment]
Confirmed, I'm still seeing this unexpected warning in 3.10 (VS 16.10) at the moment.

**Comment by RikkiGibson** [comment]
Also occurs with `class?` constraint and in the absence of constraints.

**Comment by jcouv** [comment]
Fixed for 17.0 (p1 or p2)

**Comment by jcouv** [comment]
Also backported fix to 16.11
